Logger: Enhance logging for SmtBlockComparator

PLukas2018 opened the pull request:
Providing more logs for `SmtBlockComparator` (`sequential-alu-ops`) to make it easier to debug problems.

Based on level following things are logged:
- 1+: Start of using `sequential-alu-ops` pattern, success/failure/error of the pattern
- 3+: SMT formula used for instructions comparison

This PR also:
- disables logging when trying to find end of a snippet,
- creates macros for checking what type of debug level is on.

The levels at which the logs are produced are place for discussion.
 
---
# Examples of outputs
Follows snippets of outputs of the logger run using #411.

## Examples for https://github.com/diffkemp/diffkemp/commit/a7be53eef62fafa7edf0d08b618231acd26e90d0 problem

### SIMPLL_VERBOSITY=1
#### Before
```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.60 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
SMT solver ran out of time/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (568 ms)
[----------] 1 test from DFCSmtTest (568 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (568 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```
#### Now
```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.57 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
Trying to use sequential-alu-ops pattern 
  ERROR: SMT solver ran out of time
/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (527 ms)
[----------] 1 test from DFCSmtTest (527 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (527 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```

### SIMPLL_VERBOSITY=2
#### Before
```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.57 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
L instruction:   %3 = add i8 %2, 3
R instruction:   %4 = mul i8 %1, 3
L instruction:   %3 = add i8 %2, 3
R instruction:   %5 = add i8 %3, %4
  L value:   %2 = add i8 2, 0
  R value:   %3 = mul i8 %1, %2
L instruction:   %3 = add i8 %2, 3
R instruction:   ret i8 %5
L instruction:   %4 = mul i8 %1, %3
R instruction:   %3 = mul i8 %1, %2
  L value:   %3 = add i8 %2, 3
  R value:   %2 = add i8 2, 0
L instruction:   %4 = mul i8 %1, %3
R instruction:   %4 = mul i8 %1, 3
  L value:   %3 = add i8 %2, 3
  R value: i8 3
L instruction:   %4 = mul i8 %1, %3
R instruction:   %5 = add i8 %3, %4
L instruction:   %4 = mul i8 %1, %3
R instruction:   ret i8 %5
L instruction:   ret i8 %4
R instruction:   %3 = mul i8 %1, %2
L instruction:   ret i8 %4
R instruction:   %4 = mul i8 %1, 3
L instruction:   ret i8 %4
R instruction:   %5 = add i8 %3, %4
SMT solver ran out of time/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (527 ms)
[----------] 1 test from DFCSmtTest (527 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (527 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```
#### Now

```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.57 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
Trying to use sequential-alu-ops pattern 
  ERROR: SMT solver ran out of time
/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (528 ms)
[----------] 1 test from DFCSmtTest (528 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (528 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```

### SIMPLL_VERBOSITY=3
#### Before
```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.56 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
L instruction:   %1 = add i8 1, 0
R instruction:   %1 = add i8 1, 0
L instruction:   %2 = add i8 2, 0
R instruction:   %2 = add i8 2, 0
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
Trying to find possible relocation:
  L instruction:   %3 = add i8 %2, 3
  R instruction:   %4 = mul i8 %1, 3
  L instruction:   %3 = add i8 %2, 3
  R instruction:   %5 = add i8 %3, %4
    L value:   %2 = add i8 2, 0
    R value:   %3 = mul i8 %1, %2
  L instruction:   %3 = add i8 %2, 3
  R instruction:   ret i8 %5
  L instruction:   %4 = mul i8 %1, %3
  R instruction:   %3 = mul i8 %1, %2
    L value:   %3 = add i8 %2, 3
    R value:   %2 = add i8 2, 0
  L instruction:   ret i8 %4
  R instruction:   %3 = mul i8 %1, %2
  FAILURE: Possible relocation was NOT found
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %3 = add i8 %2, 3
R instruction:   %4 = mul i8 %1, 3
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %3 = add i8 %2, 3
R instruction:   %5 = add i8 %3, %4
  L value:   %2 = add i8 2, 0
  R value:   %3 = mul i8 %1, %2
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %3 = add i8 %2, 3
R instruction:   ret i8 %5
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %4 = mul i8 %1, %3
R instruction:   %3 = mul i8 %1, %2
  L value:   %3 = add i8 %2, 3
  R value:   %2 = add i8 2, 0
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %4 = mul i8 %1, %3
R instruction:   %4 = mul i8 %1, 3
  L value:   %3 = add i8 %2, 3
  R value: i8 3
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %4 = mul i8 %1, %3
R instruction:   %5 = add i8 %3, %4
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   %4 = mul i8 %1, %3
R instruction:   ret i8 %5
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   ret i8 %4
R instruction:   %3 = mul i8 %1, %2
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   ret i8 %4
R instruction:   %4 = mul i8 %1, 3
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   ret i8 %4
R instruction:   %5 = add i8 %3, %4
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
L instruction:   ret i8 %4
R instruction:   ret i8 %5
SMT solver ran out of time/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (527 ms)
[----------] 1 test from DFCSmtTest (527 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (527 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```
#### Now

```txt
59/84 Test #59: DFCSmtTest.SmtDistributive .............................................***Failed    0.64 sec
Note: Google Test filter = DFCSmtTest.SmtDistributive
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from DFCSmtTest
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
L instruction:   %1 = add i8 1, 0
R instruction:   %1 = add i8 1, 0
L instruction:   %2 = add i8 2, 0
R instruction:   %2 = add i8 2, 0
L instruction:   %3 = add i8 %2, 3
R instruction:   %3 = mul i8 %1, %2
Scope for macro not found
Scope for macro not found
Scope for macro not found
Scope for macro not found
Trying to find possible relocation:
  L instruction:   %3 = add i8 %2, 3
  R instruction:   %4 = mul i8 %1, 3
  L instruction:   %3 = add i8 %2, 3
  R instruction:   %5 = add i8 %3, %4
    L value:   %2 = add i8 2, 0
    R value:   %3 = mul i8 %1, %2
  L instruction:   %3 = add i8 %2, 3
  R instruction:   ret i8 %5
  L instruction:   %4 = mul i8 %1, %3
  R instruction:   %3 = mul i8 %1, %2
    L value:   %3 = add i8 %2, 3
    R value:   %2 = add i8 2, 0
  L instruction:   ret i8 %4
  R instruction:   %3 = mul i8 %1, %2
  FAILURE: Possible relocation was NOT found
Trying to use sequential-alu-ops pattern 
  SMT formula:
    ; 
    (set-info :status unknown)
    (declare-fun r_var_0x5d7380 () (_ BitVec 8))
    (declare-fun l_var_0x5d4a80 () (_ BitVec 8))
    (declare-fun l_var_0x5cf3d0 () (_ BitVec 8))
    (declare-fun r_var_0x5d6cb0 () (_ BitVec 8))
    (declare-fun l_var_0x5a8060 () (_ BitVec 8))
    (declare-fun l_var_0x5d61c0 () (_ BitVec 8))
    (declare-fun r_var_0x5d7440 () (_ BitVec 8))
    (declare-fun r_var_0x5d7500 () (_ BitVec 8))
    (declare-fun r_var_0x5d75e0 () (_ BitVec 8))
    (assert
     (= l_var_0x5d4a80 r_var_0x5d7380))
    (assert
     (= l_var_0x5cf3d0 (bvadd l_var_0x5d4a80 (_ bv3 8))))
    (assert
     (= l_var_0x5a8060 r_var_0x5d6cb0))
    (assert
     (= l_var_0x5d61c0 (bvmul l_var_0x5a8060 l_var_0x5cf3d0)))
    (assert
     (= r_var_0x5d7440 (bvmul r_var_0x5d6cb0 r_var_0x5d7380)))
    (assert
     (= r_var_0x5d7500 (bvmul r_var_0x5d6cb0 (_ bv3 8))))
    (assert
     (= r_var_0x5d75e0 (bvadd r_var_0x5d7440 r_var_0x5d7500)))
    (assert
     (not (and true (= l_var_0x5d61c0 r_var_0x5d75e0))))
    (check-sat)
  ERROR: SMT solver ran out of time
/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1098: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: -1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (596 ms)
[----------] 1 test from DFCSmtTest (596 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (596 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] DFCSmtTest.SmtDistributive

 1 FAILED TEST
```

## Example for successful SMT comparison

```txt
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
Trying to use sequential-alu-ops pattern 
  SUCCESS
[       OK ] DFCSmtTest.SmtDistributive (2580 ms)
```

## Example for unsuccessful SMT compare
```
[ RUN      ] DFCSmtTest.SmtDistributive
Analysing debugging information...
Analysing debugging information...
Trying to find possible relocation:
  FAILURE: Possible relocation was NOT found
Trying to use sequential-alu-ops pattern 
  FAILURE
/home/lukas/diffkemp-org/diffkemp/tests/unit_tests/simpll/DFCLlvmIrTest.cpp:1099: Failure
Expected equality of these values:
  DiffComp->compare()
    Which is: 1
  0

[  FAILED  ] DFCSmtTest.SmtDistributive (24 ms)
```

DanielKriz reviewed (approved):
I have found only minor things, overall hte PR is good.

LGTM.

PLukas2018 commented:
> Wait, is the fix of the timeout related to the logging of SMT? I was thinking that you were adding logging in order to find the reason of the failing test...
> 
> If so, I would split it into two comments:
> 
>     1. Addition of logging
> 
>     2. Fix of the failing test

No, the failing test was fixed and is already merged into the master. I increased the timeout again because the test was failing again when logging was enabled. I thought it was because of the logging, but we start counting the time after the logs. Maybe the compiler somehow reorders the instructions? I will look into it.

PLukas2018 commented:
> I increased the timeout again because the test was failing again when logging was enabled. I thought it was because of the logging, but we start counting the time after the logs. Maybe the compiler somehow reorders the instructions? I will look into it.

So I rechecked it, and the timeout is not necessary to increase. I removed it. I think I originally had the logging below [the start of the timer](https://github.com/PLukas2018/diffkemp/blob/59fbdb1478535bbb480358f21a4c5dd9acebe25c/diffkemp/simpll/SmtBlockComparator.cpp#L692), so I needed to increase the timeout, then I moved it above so it is not necessary to increase the timeout anymore.

viktormalik commented on diffkemp/simpll/Logger.h:
```diff
@@ -106,6 +106,10 @@
 #define LOGGER_BASE_LEVEL DEBUG_SIMPLL_VERBOSE
 #define LOGGER_FORCE_LEVEL DEBUG_SIMPLL_VERBOSE_EXTRA
 
+// Checks if given logger level is turned on
+#define IS_LOG_VERBOSE_ON() isCurrentDebugType(LOGGER_BASE_LEVEL)
+#define IS_LOG_VERBOSE_EXTRA_ON() isCurrentDebugType(LOGGER_FORCE_LEVEL)
```
Can we put these into a separate commit?

PLukas2018 replied:
Done, thanks.
